### ISSUE TITLE
Adapt -except- rules and the indentation rule to use optional options objects

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#207]
* [x] `rule-non-nested-empty-line-before`
* [x] `rule-nested-empty-line-before`
* [x] `at-rule-empty-line-before`
* [x] `comment-empty-line-before`
* [x] `indentation`

Same as https://github.com/stylelint/stylelint/issues/185 but for at-rules.

**Comment by davidtheclark** [comment]
So now we run into a problem with the simple keyword option convention.

This rule already has the special `"always-except-blockless-group"` option. If we want to add an `"always-except-first"` option, as well, it seems we'd almost have to add a `"always-except-first-or-blockless-group"` option in case the user wanted both exceptions ... and that of course could continue indefinitely.

**Comment by davidtheclark** [comment]
Can anybody see a way out of this? Or does it mean that we should have the first (I think) *object* of options? Something like ...

```
'at-rule-empty-line-before`: ['always', {
  ignoreBlocklessGroup: true,
  ignoreFirst: true
}]
```

**Comment by davidtheclark** [comment]
I'm reminded we already have an object for `indentation`.

As an alternative to the above to avoid `ignore*` options, we could do

```js
'at-rule-empty-line-before`: [2, {
  default: "always",
  blocklessGroups: "never",
  first: "always"
}]
```

I don't know. Maybe.

**Comment by jeddy3** [comment]
I realised the same thing yesterday.

I'm also not sure what the right solution is. Although, I think eslint's approach of an *additional and optional* option object might give us the necessary balance between flexibility and keeping things easy for the user (i.e. them not having to think about the optional options most of the time).

So, `at-rule-empty-line-before` might look something like:

```js
"at-rule-empty-line-before": [2, "always", {
  except: [
    "blockless-group",
    "first",
  ]
}]
```

"Except" means: *enforce the opposite* of the 1st option i.e. *never* an empty line before the first rule or rules in a blockless-group.

If that's too stringent we could also provide an `ignore` array within the optional options object e.g.

```js
"at-rule-empty-line-before": [2, "always", {
  except: ["blockless-group",],
  ignore: ["first"]
}]
```

i.e. always an empty line, except never in a blockless-group, and ignore the first node (so anything goes for that one). 

Another use of `ignore` would be in `comment-empty-line-before` e.g.

```js
"comment-empty-line-before": [2, "always", {
  ignore: ["inline"]
}]
```

We currently have only a handful of rules using these special options, so this seems like the ideal time to have this discussion :) As you pointed out, `indentation` is one of them. Perhaps that could work as follows:

Indent everything by a tab:

```js
{
  "indentation": [ 2, "tab" ],
}
```

Indent everything by 4 spaces:

```js
{
  "indentation": [ 2, 4 ],
}
```

Indent everything by a tab, except values (this is, as before, enforce the opposite i.e. check that there is *no* indentation before values).

```js
{
  "indentation": [ 2, "tab", {
    "except": ["value"]
  }, ],
}
```

Kind just throwing ideas out there, but does this seem intuitive/easiest from a user's perspective while being flexible enough for us to add exceptions to our current and future rules?

**Comment by davidtheclark** [comment]
Quick question: Is that except=reverse convention in ESLint? Could you point to an example? I tried searching the codebase real quick but didn't come up with something that worked like this -- I did see some `"except-*"` keywords...

**Comment by jeddy3** [comment]
> Quick question: Is that except=reverse convention in ESLint?

Oh... I was referring to just the *optional options object* bit of eslint, but I did just find this example of "except=reverse":
http://eslint.org/docs/rules/space-in-parens#exceptions

There seems to be a quite a few different conventions in use:

As you pointed out, there's also `"except-*"` keywords... like here:
http://eslint.org/docs/rules/no-cond-assign.html

There's also a couple of `*-no-*` rules that except optional objects e.g.:
http://eslint.org/docs/rules/no-multiple-empty-lines

Perhaps this flexibility is needed?

**Comment by davidtheclark** [comment]
I think I like the `except` and `ignore` idea. That would provide *a lot* of flexibility to accommodate future requests.

So how about this if you don't care whether blockless groups are separated by an empty line and you definitely want *no* line before the first nested at-rule:

```js
'at-rule-empty-line-before`: [2, "always", {
  except: [ 'first' ],
  ignore: [ 'blocklessGroups' ],
}],
```

Good?

**Comment by jeddy3** [comment]
> Good?

Yeap. That's right.

It looks like only three rules of affected by this change:

* `rule-nested-empty-line-before`
* `at-rule-empty-line-before`
* `comment-empty-line-before`

I'm happy to go down this route if you are? :)

**Comment by davidtheclark** [comment]
I'm in.

**Comment by jeddy3** [comment]
I'm going to try to adapt `at-rule-empty-line-before` now.

**Comment by jeddy3** [comment]
@davidtheclark I'm OK with adapting the rules, but this change also touches on `ruleTester` util. Are you able to adapt `ruleTester` to take an optional options object for me as I'm not too sure on the best approach?

**Comment by davidtheclark** [comment]
Yes. I can do it after work this evening probably.

**Comment by jeddy3** [comment]
Thank you! :)

**Comment by jeddy3** [comment]
Closed in #220
